### PR TITLE
Add a method to construct a RelativePath to get from one AbsolutePath to another

### PR DESCRIPTION
**Pull request by abertelrud**
This is needed by some parts of the SwiftPM source code in order to switch them over to AbsolutePath and RelativePath.  There was such a method for strings, but this makes it type-safe.

### CODE REVIEW

File: Sources/Basic/Path.swift [hunk 1]
```diff
@@ -327,6 +327,51 @@ private struct PathImpl {
...
+    /// returned path will begin with one or more `..` path components.
+    ///
+    /// Because both paths are absolute, they always have a common ancestor
+    /// (the root path, if nothing else).  Therefore, any path can be made
+    /// relative to any other path by using a sufficient number of `..` path
+    /// components.
+    ///
+    /// This method is strictly syntactic and does not access the file system
+    /// in any way.  Therefore, it does not take symbolic links into account.
+    public func relative(to base: AbsolutePath) -> RelativePath {
```
Comment (ddunbar) [inline Sources/Basic/Path.swift]: I wonder if we should try and pick a more Swift3 style name, like `madeRelative(to:)` (not sure that is right, but maybe conveys the idea)?

File: Sources/Basic/Path.swift [hunk 2]
```diff
@@ -327,6 +327,51 @@ private struct PathImpl {
...
+                newPathComps = newPathComps.dropFirst()
+                newBaseComps = newBaseComps.dropFirst()
+            }
+            // Now construct a path consisting of as many `..`s as are in the
+            // `newBaseComps` followed by what remains in `newPathComps`.
+            var relComps = Array(repeating: "..", count: newBaseComps.count)
+            relComps.append(contentsOf: newPathComps)
+            return RelativePath(relComps.joined(separator: "/"))
+        }
+    }
```
Comment (ddunbar) [inline Sources/Basic/Path.swift]: Can this method have `assert(base.join(result) == self)`?

**Comment by ddunbar** [comment]
LGTM, with minor comments inline

File: Sources/Basic/Path.swift (hunk 1 above)
Comment (abertelrud) [inline Sources/Basic/Path.swift]: I'm on the fence about this, and the Swift 3 style doesn't seem entirely consistent.  For example, while there are `joined()`, `separated()`, etc there are also `filter()`, `map()`, `dropLast()` etc instead of `filtered()`, `mapped()`, `withLastDropped()`, etc.  There may be some pattern to those names that I'm missing.

Comment (ddunbar) [inline Sources/Basic/Path.swift]: Ok, let's stick with existing name then.

Comment (abertelrud) [inline Sources/Basic/Path.swift]: Sounds good.  I'm totally open to change if there's a compelling alternative that makes the call sites look good (and ideally not be too wordy).

File: Sources/Basic/Path.swift (hunk 2 above)
Comment (abertelrud) [inline Sources/Basic/Path.swift]: Sure, I'll add that (now that I've confirmed that `assert()` is omitted from release builds).

File: Sources/Basic/Path.swift (hunk 1 above)
Comment (abertelrud) [inline Sources/Basic/Path.swift]: BTW, this comment made me think that we should rename `join()` to `joined()`.  Personally I hate both, since `absPath + relPath` seems pretty clear to me (i.e. an operator).  I would not go so far as to suggest a `-` operator for getting a relative path, however, though there would be a certain logic to it.

Comment (ddunbar) [inline Sources/Basic/Path.swift]: I agree `joined` seems better for Swift3.

Hate seems a strong word... `+` usually maps to some other named operation (not "plus"), so the name for this should be whatever name you would want to use there. I have no objections to a better word than "joined", but there should be a method in any case even if we wanted the operator.

Comment (abertelrud) [inline Sources/Basic/Path.swift]: Yes, that was a bit of hyperbole.  Strongly dislike, then, because the actual operation really seems more like appending a second thing to a first thing, whereas joining (in the sense in which our APIs use it) is to combine all the elements of a sequence.

I actually think that the word "append" iis more appropriate here, though that would of course be for a mutating method.

I definitely agree that an operator, if we have one, would be just an alias for whatever name we choose.

Comment (abertelrud) [inline Sources/Basic/Path.swift]: Regarding mutating:  really, the path classes should work both in mutable and immutable form, but that is a topic of another enhancement somewhere down the road.

Comment (ddunbar) [inline Sources/Basic/Path.swift]: I agree `append` (or `appending`, if we think that is more Swift3) is better.

Comment (abertelrud) [inline Sources/Basic/Path.swift]: Perfect.  `appending` it is.  I agree completely about the mutable vs immutable forms.

I don't think I had put into words before why `join` felt so wrong as a name, but it's pretty obvious now.  I'll change it to `appending`.  Does that affect your opinion about the best name for `relative`?  It doesn't feel completely right but I can't think of anything better (`madeRelative` is a bit to awkward, I think).

**Comment by abertelrud** [comment]
@swift-ci please test

File: Sources/Basic/Path.swift (hunk 1 above)
Comment (ddunbar) [inline Sources/Basic/Path.swift]: I agree, I can't think of a better name now.
